### ISSUE TITLE
Terraform crashed when importing inputs contain integer

### ISSUE THREAD

**Issue by thaiphv** [hashicorp/terraform#17678]
### Terraform Version
```
0.11.5
```

### Terraform Configuration Files
```hcl
data "terraform_remote_state" "backbone" {
  backend = "local"

  config {
    path = "${path.cwd}/../../../../backbone/environments/prod/foundation/terraform.tfstate"
  }
}
```

### Debug Output
https://gist.github.com/thaiphv/66e0def2f5953556e75edbfe807d3585

### Crash Output
https://gist.github.com/thaiphv/25b806f0051c7c5a53a3e13355174ef0

### Expected Behavior
`terraform plan` should have produced an execution plan

### Actual Behavior
It crashed
I manually edited the state file to quotes all the  integer outputs and I was able to produce the plan following the same step.

### Steps to Reproduce
1. `cd my_module`
2. `terraform init`
3. `terraform plan`

**Comment by jbardin** [comment]
Hi @thaiphv,

Thanks for filing the issue!

From the stack trace it looks like you have a float value nested in a list, in a map. Floats are not currently a supported data type, but it seems that the validation misses them when they are in a nested data structure. 

Also note that nested lists and maps are not fully supported, and can fail in certain situations, but that is not the cause of the crash here.

**Comment by thaiphv** [comment]
Hi @jbardin,

The output actually contained a map of arrays of integers.

```lang=hcl
output "asns" {
  value = {
    ap-southeast-1 = [60000, 60001]
  }
}
```

**Comment by jbardin** [comment]
@thaiphv,

Yes, that has the same effect, since we can't declare types for nested data structures, they are deserialized as float64 by default. If you were to put a  numeric value directly in the output, validation would fail with the error: `output asns is not a valid type (int)`.

**Comment by apparentlymart** [comment]
Hi @thaiphv! Sorry for the long silence here.

We've recently merged in to master support for outputs of any type, whereas before (as James mentioned) we only supported simple types.

I verified with the v0.12.0-alpha1 prerelease build using the output example in your most recent comment:

```hcl
output "asns" {
  value = {
    ap-southeast-1 = [60000, 60001]
  }
}
```

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

asns = {
  "ap-southeast-1" = [
    60000,
    60001,
  ]
}
```

This is now stored in the state with full type information, allowing `terraform_remote_state` to properly decode these values back into the same types that were given in the producing configuration:

```json
{
  "version": 4,
  "terraform_version": "0.12.0",
  "serial": 1,
  "lineage": "1104157e-1db7-22d7-b67f-e47a8c6654e7",
  "outputs": {
    "asns": {
      "value": {
        "ap-southeast-1": [
          60000,
          60001
        ]
      },
      "type": [
        "object",
        {
          "ap-southeast-1": [
            "tuple",
            [
              "number",
              "number"
            ]
          ]
        }
      ]
    }
  },
  "resources": []
}
```

"object" and "tuple" are the two new type kinds we've added to represent mappings and sequences where not all of the elements are of the same type. In this case it could also have been represented as a map of lists of numbers, but since Terraform can make these conversions automatically when needed it will always use the most general form in situations like outputs where there's no explicit type given.

These fixes are already merged and will be included in the forthcoming v0.12.0 final release, and so I'm going to close this now. Thanks for reporting this, and sorry for the delay in resolving it.
